engine/src/main/java/org/kigalisim/engine/number/UnitConverter.java

```diff
@@ -0,0 +1,600 @@
+ *
+ * <p>This class provides functionality to convert between different units used in the 
+ * simulation engine, including volume, population, consumption, and time-based units.
+ * It uses BigDecimal for numerical precision and stability.</p>
+ *
+ * @license BSD-3-Clause
+ */
+
+package org.kigalisim.engine.number;
+
+import java.math.BigDecimal;
+import java.math.MathContext;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * Object simplifying conversion between units.
+ *
+ * <p>This class handles unit conversions within the engine, supporting conversions
+ * between volume units (kg, mt), population units (unit, units), consumption units
+ * (tCO2e, kwh), time units (year, years), and percentage units (%).</p>
+ */
+public class UnitConverter {
+  
+  // Configuration constants
+  private static final boolean CONVERT_ZERO_NOOP = true;
+  private static final boolean ZERO_EMPTY_VOLUME_INTENSITY = false;
+  
+  // Math context for BigDecimal operations
+  private static final MathContext MATH_CONTEXT = MathContext.DECIMAL128;
+  
+  // Conversion factors
+  private static final BigDecimal KG_TO_MT_FACTOR = new BigDecimal("1000");
+  private static final BigDecimal PERCENT_FACTOR = new BigDecimal("100");
+  
+  private final StateGetter stateGetter;
+  
+  /**
+   * Create a new unit converter.
+   *
+   * @param stateGetter Object allowing access to engine state as needed for unit conversion
+   */
+  public UnitConverter(StateGetter stateGetter) {
+    this.stateGetter = stateGetter;
+  }
+  
+  /**
+   * Convert a number to new units.
+   *
+   * @param source The EngineNumber to convert
+   * @param destinationUnits The units to which source should be converted
+   * @return The converted EngineNumber
+   */
+  public EngineNumber convert(EngineNumber source, String destinationUnits) {
+    if (source.getUnits().equals(destinationUnits)) {
+      return source;
+    }
+    
+    if (CONVERT_ZERO_NOOP && source.getValue().compareTo(BigDecimal.ZERO) == 0) {
+      return new EngineNumber(BigDecimal.ZERO, destinationUnits);
+    }
+    
+    String[] sourceUnitPieces = source.getUnits().split(" / ");
+    boolean sourceHasDenominator = sourceUnitPieces.length > 1;
+    String sourceDenominatorUnits = sourceHasDenominator ? sourceUnitPieces[1] : "";
+    
+    String[] destinationUnitPieces = destinationUnits.split(" / ");
+    boolean destHasDenominator = destinationUnitPieces.length > 1;
+    String destinationDenominatorUnits = destHasDenominator ? destinationUnitPieces[1] : "";
+    
+    String sourceNumeratorUnits = sourceUnitPieces[0];
+    String destinationNumeratorUnits = destinationUnitPieces[0];
+    boolean differentDenominator = !destinationDenominatorUnits.equals(sourceDenominatorUnits);
+    boolean sameDenominator = !differentDenominator;
+    
+    // Create strategy maps
+    Map<String, Function<EngineNumber, EngineNumber>> numeratorStrategy = 
```

Comment:
Instead of function, let’s create a `UnitConverterStrategy` (please create this in a new file `UnitConverterStrategy.java` that formalizes this interface.